src/DocBlock/Annotation.php

```diff
@@ -16,10 +16,39 @@
  * This represents an entire annotation from a docblock.
  *
  * @author Graham Campbell <graham@alt-three.com>
+ * @author Dariusz Rumiński <dariusz.ruminski@gmail.com>
  */
 class Annotation
 {
     /**
+     * Regex to match any types, shall be use with `x` modifier.
+     *
+     * @internal
+     */
+    const REGEX_TYPES = '
+    # <simple> is any non-array, non-generic, non-alternated type, eg `int` or `\Foo`
+    # <generic> is generic collection type, like `array<string, int>`, `Collection<Item>` and more complex like `Collection<int, \null|SubCollection<string>>`
+    # <type> is <simple>, <simple>[] or <generic> type, like `int`, `bool[]` or `Collection<ItemKey, ItemVal>`
+    # <types> is one or more types alternated via `|`, like `int|bool[]|Collection<ItemKey, ItemVal>`
+    (?<types>
+        (?<type>
+            (?:(?&simple)(?:\[\])?)
+            |
+            (?<generic>
+                (?<simple>
+                    [\\\\@$]?[\\\\\w]+
```

Comment:
No need to match `\` in the first character class as it will be matched by the second.